File: programme/quick-sort/quick-sort.c
// Write a programme for quick sort
#include <stdio.h>

// Utility function to swap array elements
void swap(int a[], int i, int j)
{
  int t = a[i];
  a[i] = a[j];
  a[j] = t;
}

int partition(int arr[], int start, int end)
{
  // initialise last element as pivot
  int x = arr[end];
  int i = start - 1, temp;

  for (int j = start; j <= end - 1; j++)
  {
    // if current element is smaller than pivot
    if (arr[j] < x)
    {
      // increment index of smaller element
      i++;
      swap(arr, i, j);
    }
  }
  swap(arr, i + 1, end);

  return i + 1;
}

void qs(int a[], int start, int end)
{
  if (start > end)
    return;

  int pivot = partition(a, start, end);

  // Sorts elements before the pivot and after the pivot separately
  qs(a, start, pivot - 1);
  qs(a, pivot + 1, end);
}

void quickSort(int input[], int size)
{

  qs(input, 0, size - 1);

  // Prints the sorted array
  for (int i = 0; i < size; i++)
    printf("%d ", input[i]);

  printf("\n");
}

int main()
{
  int a[] = {10, 80, 30, 90, 40, 50, 70};
  printf("Input  : ");
  for (int i = 0; i < 7; i++)
    printf("%d ", a[i]);
  printf("\n");

  int n = sizeof(a) / sizeof(a[0]);
  printf("Output : ");
  quickSort(a, n);

  return 0;
}

File: programme/quick-sort/quick-sort.js
function swap(arr, i, j){
    const temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
}


function partition(arr, left, right) {
    const temp = arr[left]
    let swapIndex = left

    for (let i = left + 1; i <= right; i++) {
      if (temp > arr[i]) {
        swapIndex++
        if (i !== swapIndex) 
          swap(arr, swapIndex, i)
      }
    }

    if (swapIndex !== left) 
      swap(arr, swapIndex, left)

    return swapIndex
  }


  function quickSort(arr, left = 0, right = arr.length - 1) {
    if (left >= right) return
    
    let pivot = partition(arr, left, right)

    quickSort(arr, left, pivot - 1)
    quickSort(arr, pivot + 1, right)
    return arr
  }



let a = [5, 4, 1, 6, 2]

console.log("Before Quick Sort: " + a)
quickSort(a)
console.log("After Quick Sort:  " + a)
